go/tools/gopackagesdriver/main.go

```diff
@@ -45,77 +49,64 @@ type driverResponse struct {
+	bazelBin              = getenvDefault("GOPACKAGESDRIVER_BAZEL", "bazel")
+	bazelFlags            = strings.Fields(os.Getenv("GOPACKAGESDRIVER_BAZEL_FLAGS"))
+	bazelQueryFlags       = strings.Fields(os.Getenv("GOPACKAGESDRIVER_BAZEL_QUERY_FLAGS"))
+	bazelQueryScope       = getenvDefault("GOPACKAGESDRIVER_BAZEL_QUERY_SCOPE", "")
+	bazelBuildFlags       = strings.Fields(os.Getenv("GOPACKAGESDRIVER_BAZEL_BUILD_FLAGS"))
+	workspaceRoot         = os.Getenv("BUILD_WORKSPACE_DIRECTORY")
+	emptyResponse         = &driverResponse{
+		NotHandled: false,
+		Sizes:      types.SizesFor("gc", "amd64").(*types.StdSizes),
+		Roots:      []string{},
+		Packages:   []*FlatPackage{},
 	}
-	return defaultValue
-}
-
-func signalContext(parentCtx context.Context, signals ...os.Signal) (ctx context.Context, stop context.CancelFunc) {
-	ctx, cancel := context.WithCancel(parentCtx)
-	ch := make(chan os.Signal, 1)
-	go func() {
-		select {
-		case <-ch:
-			cancel()
-		case <-ctx.Done():
-		}
-	}()
-	signal.Notify(ch, signals...)
-
-	return ctx, cancel
-}
+)
 
-func run() error {
+func run() (*driverResponse, error) {
 	ctx, cancel := signalContext(context.Background(), os.Interrupt)
 	defer cancel()
 
+	queries := os.Args[1:]
+
 	request, err := ReadDriverRequest(os.Stdin)
 	if err != nil {
-		return fmt.Errorf("unable to read request: %w", err)
+		return emptyResponse, fmt.Errorf("unable to read request: %w", err)
 	}
 
 	bazel, err := NewBazel(ctx, bazelBin, workspaceRoot)
 	if err != nil {
-		return fmt.Errorf("unable to create bazel instance: %w", err)
+		return emptyResponse, fmt.Errorf("unable to create bazel instance: %w", err)
 	}
 
-	bazelJsonBuilder, err := NewBazelJSONBuilder(bazel, targetsQueryStr, targetsTagFilters, targets)
+	bazelJsonBuilder, err := NewBazelJSONBuilder(bazel, queries...)
 	if err != nil {
-		return fmt.Errorf("unable to build JSON files: %w", err)
+		return emptyResponse, fmt.Errorf("unable to build JSON files: %w", err)
 	}
 
 	jsonFiles, err := bazelJsonBuilder.Build(ctx, request.Mode)
 	if err != nil {
-		return fmt.Errorf("unable to build JSON files: %w", err)
+		return emptyResponse, fmt.Errorf("unable to build JSON files: %w", err)
 	}
 
 	driver, err := NewJSONPackagesDriver(jsonFiles, bazelJsonBuilder.PathResolver())
 	if err != nil {
-		return fmt.Errorf("unable to load JSON files: %w", err)
+		return emptyResponse, fmt.Errorf("unable to load JSON files: %w", err)
 	}
 
-	response := driver.Match(os.Args[1:]...)
-	// return nil
-
-	if err := json.NewEncoder(os.Stdout).Encode(response); err != nil {
-		return fmt.Errorf("unable to encode response: %w", err)
-	}
-
-	return nil
+	return driver.Match(queries...), nil
 }
 
 func main() {
-	if err := run(); err != nil {
+	response, err := run()
+	if err := json.NewEncoder(os.Stdout).Encode(response); err != nil {
+		fmt.Fprintf(os.Stderr, "unable to encode response: %v", err)
+	}
+	if err != nil {
 		fmt.Fprintf(os.Stderr, "error: %v", err)
-		os.Exit(1)
+		os.Exit(0)
```

Comment:
Why exiting with 0?

Reply:
because gopls will check the packages driver exit code, and if there is an error, and fall back to `go list`: https://github.com/golang/tools/blob/8c34cc9cafff8f13c3f3ed95a11de309752c1c60/go/packages/external.go#L88-L90